presto-main/src/main/java/com/facebook/presto/operator/exchange/LocalExchange.java

```diff
@@ -121,8 +131,63 @@ else if (partitioning.equals(FIXED_PASSTHROUGH_DISTRIBUTION)) {
             };
         }
         else {
-            throw new IllegalArgumentException("Unsupported local exchange partitioning " + partitioning);
+            // partitioned exchange
+            exchangerSupplier = () -> new PartitioningExchanger(
+                    buffers,
+                    memoryManager,
+                    createPartitionFunction(
+                            partitioningProviderManager,
+                            session,
+                            partitioning,
+                            bufferCount,
+                            partitioningChannelTypes,
+                            partitionHashChannel.isPresent()),
+                    partitionChannels,
+                    partitionHashChannel);
+        }
+    }
+
+    private static PartitionFunction createPartitionFunction(
+            PartitioningProviderManager partitioningProviderManager,
+            Session session,
+            PartitioningHandle partitioning,
+            int partitionCount,
+            List<Type> partitioningChannelTypes,
+            boolean isHashPrecomputed)
+    {
+        if (partitioning.getConnectorHandle() instanceof SystemPartitioningHandle) {
+            HashGenerator hashGenerator;
+            if (isHashPrecomputed) {
+                hashGenerator = new PrecomputedHashGenerator(0);
+            }
+            else {
+                hashGenerator = new InterpretedHashGenerator(partitioningChannelTypes, IntStream.range(0, partitioningChannelTypes.size()).toArray());
+            }
+            return new LocalPartitionGenerator(hashGenerator, partitionCount);
+        }
+
+        ConnectorNodePartitioningProvider partitioningProvider = partitioningProviderManager.getPartitioningProvider(partitioning.getConnectorId().get());
```

Comment:
I can see why we cannot reuse `NodePartitioningManager#getPartitionFunction` (as explained in https://github.com/prestodb/presto/pull/13294#discussion_r320882561) .

Still, might be helpful to refactor some common logic into a method that returns `BucketFunction`. We can do this refactor later though.